source/llassetgen/source/DistanceTransform.cpp

```diff
@@ -1,300 +1,166 @@
-#include <algorithm>
-#include <limits>
-
-/*
- * If we sort png.h below the freetype includes, an compile error will be
- * generated warning about multiple includes of setjmp (on some version of
- * libpng). Because we don't use the setjmp functionality of freetype, this is
- * not applicable in this case. Unfortunately, the compile error can only be
- * fixed by editing png.h or reordering the includes.
- *
- * See https://sourceforge.net/p/enlightenment/mailman/message/12595025/ for
- * some discussion on this topic.
- */
-// clang-format off
-#include <png.h> // NOLINT
-#include <ft2build.h>
-#include FT_FREETYPE_H
-// clang-format on
+#include <iostream>
+#include <cassert>
+#include <cmath>
 
 #include <llassetgen/DistanceTransform.h>
 
-template <typename T>
-T square(T value) {
-    return value * value;
-}
-
-template <typename T>
-T clamp(T val, T min, T max) {
-    return std::max(min, std::min(max, val));
-}
-
 namespace llassetgen {
-    void DistanceTransform::resetInput(DimensionType _width, DimensionType _height, bool clear) {
-        assert(_width > 0 && _height > 0);
-        width = _width;
-        height = _height;
-        DimensionType size = (width * height + 7) / 8;
-        input.reset(new InputType[size]);
-        if (clear) memset(input.get(), 0, size);
-    }
-
-    bool DistanceTransform::inputAt(DimensionType offset) {
-        assert(offset < width * height && input.get());
-        return (input[offset / 8] >> (offset % 8)) & 1;
-    }
-
-    bool DistanceTransform::inputAt(PositionType pos) {
-        assert(pos.x < width && pos.y < height);
-        return inputAt(pos.y * width + pos.x);
-    }
-
-    bool DistanceTransform::inputAtClamped(PositionType pos) {
-        if (static_cast<int>(pos.x) < 0 || static_cast<int>(pos.y) < 0 || pos.x >= width || pos.y >= height) return 0;
-        return inputAt(pos);
-    }
-
-    void DistanceTransform::inputAt(PositionType pos, bool bit) {
-        assert(pos.x < width && pos.y < height && input.get());
-        DimensionType i = pos.y * width + pos.x;
-        InputType mask = 1 << (i % 8);
-        i /= 8;
-        if (bit)
-            input[i] |= mask;
-        else
-            input[i] &= ~mask;
-    }
-
-    DistanceTransform::OutputType& DistanceTransform::outputAt(DimensionType offset) {
-        assert(offset < width * height && output.get());
-        return output[offset];
-    }
-
-    DistanceTransform::OutputType& DistanceTransform::outputAt(PositionType pos) {
-        assert(pos.x < width && pos.y < height && output.get());
-        return output[pos.y * width + pos.x];
+    template<typename PixelType, bool flipped, bool invalidBounds>
+    PixelType DistanceTransform::getPixel(PositionType pos) {
+        const Image& image = std::is_same<PixelType, InputType>::value ? input : output;
+        if(invalidBounds && !image.isValid(pos))
+            return 0;
```

Comment:
This makes every pixel outside of the bounds black in the eyes of the distance transform. Since we render black characters on white background, distance fields have a white border around them.
![testfontatlas_rgb](https://user-images.githubusercontent.com/6051384/37316772-4f68d9f2-2661-11e8-80fc-30e1b0935025.png)

Reply:
Changing the value to 1 fixes the problem when DeadReckoning is used, though I don't know if this could break `getPixel<OutputType>`. ParabolaEnvelope creates a white border regardless of the value.

Reply:
The reason for the `TODO` at the end of the line
`(value >= std::numeric_limits<InputType>::max()/2) : value; // TODO`
was that this check should not happen in DistanceTransform but in Image.
The bit depth reduction was not implemented back then, thus this ugly solution.

Reply:
So we have to decide what black / white interpretation we want and use a consistent one and remove this discretization here.

Reply:
I see. Could you move the check to Image then?

Reply:
Assuming we choose the current b/w interpretation, do you have an idea how to fix ParabolaEnvelope?

Reply:
Doesn't TTF use 0 / black for empty and 1 / white for set?
Why would you invert the input, it only costs computation power and time...
Please keep in mind that the output can simply be inverted independent of the input by flipping black / white and min / max parameters of Image::exportPng().